pkg/chunk/repair_test.go

```diff
@@ -0,0 +1,296 @@
+func TestRecoveryHook(t *testing.T) {
+	// test variables needed to be correctly set for any recovery hook to reach the sender func
+	chunkAddr := chunktesting.GenerateTestRandomChunk().Address()
+	target := "0xedC69a0F0E81394bb08F90F89e35F93287E99dc1"
+	ctx := context.WithValue(context.Background(), api.TargetsContextKey{}, target)
+
+	// setup the sender
+	hookWasCalled := false // test variable to check if hook is called
+	testSender := func(ctx context.Context, targets trojan.Targets, topic trojan.Topic, payload []byte) (*pss.Monitor, error) {
+		hookWasCalled = true
+		return nil, nil
+	}
+
+	// create recovery hook and call it
+	recoveryHook := chunk.NewRecoveryHook(testSender)
+	if err := recoveryHook(ctx, chunkAddr); err != nil {
+		t.Fatal(err)
+	}
+
+	if hookWasCalled != true {
+		t.Fatalf("recovery hook was not called")
+	}
+}
+
+// RecoveryHookTestCase is a struct used as test cases for the TestRecoveryHookCalls func
+type RecoveryHookTestCase struct {
+	name           string
+	ctx            context.Context
+	expectsFailure bool
+}
+
+// TestRecoveryHookCalls verifies that recovery hooks are being called as expected when net store attempts to get a chunk
+func TestRecoveryHookCalls(t *testing.T) {
+	// generate test chunk, store and publisher
+	c := chunktesting.GenerateTestRandomChunk()
+	ref := c.Address()
+	p := "0xbE165fe06c03e4387F79615b7A0b79d535e8D325"
+
+	// test cases variables
+	dummyContext := context.Background() // has no publisher
+	targetContext := context.WithValue(context.Background(), api.TargetsContextKey{}, p)
+
+	for _, tc := range []RecoveryHookTestCase{
+		{
+			name:           "no targets in context",
+			ctx:            dummyContext,
+			expectsFailure: true,
+		},
+		{
+			name:           "targets set in context",
+			ctx:            targetContext,
+			expectsFailure: false,
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			hookWasCalled := make(chan bool, 1) // channel to check if hook is called
+
+			// setup recovery hook
+			testHook := func(ctx context.Context, targets trojan.Targets, topic trojan.Topic, payload []byte) (*pss.Monitor, error) {
+				hookWasCalled <- true
+				return nil, nil
+			}
+			recoverFunc := chunk.NewRecoveryHook(testHook)
+			ns := newTestNetStore(t, recoverFunc)
+
+			// fetch test chunk
+			_, err := ns.Get(tc.ctx, storage.ModeGetRequest, ref)
+			if err != nil && err != netstore.ErrRecoveryAttempt && err.Error() != "netstore retrieve chunk: get closest: no peer found" {
+				t.Fatal(err)
+			}
+
+			// checks whether the callback is invoked or the test case times out
+			select {
+			case <-hookWasCalled:
+				if !tc.expectsFailure {
+					return
+				}
+				t.Fatal("recovery hook was unexpectedly called")
+			case <-time.After(100 * time.Millisecond):
+				if tc.expectsFailure {
+					return
+				}
+				t.Fatal("recovery hook was not called when expected")
+			}
+		})
+	}
+}
+
+func TestNewRepairHandler(t *testing.T) {
```

Comment:
it is kind of hard to read/follow this test.

do you think it's possible to do this through a `for _, tc in range tcs...`-type of statement and re-use some of the code here?

Reply:
Its kinda difficult to do that since every test case needs different initialization.